packages/migrate-config/src/migrate-config.js

```diff
@@ -0,0 +1,882 @@
+ * @returns {boolean} `true` if the plugin needs the compat utility.
+ */
+function pluginNeedsCompat(pluginName) {
+	const pluginNameToTest = pluginName.includes("/")
+		? pluginName.slice(0, pluginName.indexOf("/"))
+		: pluginName;
+
+	return pluginsNeedingCompat.includes(
+		naming.normalizePackageName(pluginNameToTest, "eslint-plugin"),
+	);
+}
+
+/**
+ * Gets the name of the variable to use for the plugin. If the plugin name
+ * contains slashes or an @ symbol, it will be normalized to a camelcase name.
+ * If the name is "import" or "export", it will be prefixed with an underscore.
+ * @param {string} pluginName The name of the plugin.
+ * @returns {string} The variable name to use.
+ */
+function getPluginVariableName(pluginName) {
+	let name = pluginName.replace(/^eslint-plugin-/, "");
+
+	if (name === "import" || name === "export") {
+		return `_${name}`;
+	}
+
+	if (name.startsWith("@")) {
+		name = name.slice(1);
+	}
+
+	return camelCase(name);
+}
+
+/**
+ * Creates an initialization block for the FlatCompat utility.
+ * @param {"module"|"commonjs"} sourceType The module type to use.
+ * @returns {Array<Statement>} The AST for the initialization block.
+ */
+function getFlatCompatInit(sourceType) {
+	let init = `
+const compat = new FlatCompat({
+    baseDirectory: __dirname,
+    recommendedConfig: js.configs.recommended,
+    allConfig: js.configs.all
+});
+`;
+
+	/*
+	 * Need to calculate `__dirname` and `__filename` for ESM. Note that Recast
+	 * doesn't support `import.meta.url`, so using an uppercase "I" to allow for
+	 * parsing. We then need to replace it with the lowercase "i".
+	 */
+	if (sourceType === "module") {
+		init = `
+const __filename = fileURLToPath(Import.meta.url);
+const __dirname = path.dirname(__filename);
+
+${init}`;
+	}
+
+	const result = recast.parse(init).program.body;
+
+	// Replace uppercase "I" with lowercase "i" in "Import.meta.url"
+	if (sourceType === "module") {
+		result[0].declarations[0].init.arguments[0].object.object.name =
+			"import";
+	}
+
+	return result;
+}
+
+/**
+ * Converts a glob pattern to a format that can be used in a flat config.
+ * @param {string} pattern The glob pattern to convert.
+ * @returns {string} The converted glob pattern.
+ */
+function convertGlobPattern(pattern) {
+	const isNegated = pattern.startsWith("!");
+	const patternToTest = isNegated ? pattern.slice(1) : pattern;
+	const isDescendant = patternToTest.startsWith("**");
+	const isRoot = patternToTest.startsWith("/");
+	const isCwd = patternToTest.startsWith("./");
+	const slashIndex = patternToTest.indexOf("/");
+	const lastSlashIndex = patternToTest.lastIndexOf("/");
+	const hasSlash = slashIndex !== -1;
+	const hasMultipleSlashes = hasSlash && slashIndex !== lastSlashIndex;
+
+	// if the pattern is already in the correct format, return it
+	if (isDescendant || isRoot || isCwd || hasMultipleSlashes) {
+		return pattern;
+	}
+
+	return `${isNegated ? "!" : ""}**/${pattern}`;
+}
```

Comment:
This is to account for the fact that flat config uses minimatch with `matchBase: false` while eslintrc uses minimatch with `matchBase: true`?

It should be the same pattern if there's at least _one_ slash (or the whole pattern is `**`). For example, `src/**` should stay `src/**`, not become `**/src/**`.

https://www.npmjs.com/package/minimatch/v/3.1.2#matchbase

Reply:
Here a stackblitz example:

https://stackblitz.com/edit/stackblitz-starters-awqzxy?file=package.json

flat config with `src/**` has the same behavior as eslintrc with `src/**`.

flat config with `**/src/**` has a different behavior.

Reply:
So I think this can be simplified to:

```suggestion
function convertGlobPattern(pattern) {
	const isNegated = pattern.startsWith("!");
	const patternToTest = isNegated ? pattern.slice(1) : pattern;

	// if the pattern is already in the correct format, return it
	if (patternToTest === "**" || patternToTest.includes("/")) {
		return pattern;
	}

	return `${isNegated ? "!" : ""}**/${patternToTest}`;
}
```